src/advanced_inv.cpp

```diff
@@ -1365,111 +1428,99 @@ void advanced_inventory::display()
+                    } else if(srcarea == AIM_WORN) {
+                        std::vector<item> mv;
+                        g->u.takeoff(idx, false, &mv);
+                        std::copy(mv.begin(), mv.end(), std::back_inserter(moving_items));
+                    }
+                    int items_left = 0, moved = 0;
+                    for(auto &elem : moving_items) {
+                        assert(!elem.is_null());
+                        items_left = add_item(destarea, elem);
+                        if(items_left > 0) {
+                            // chargeback the items if adding them failed
+                            add_item(srcarea, elem, items_left);
+                        } else {
+                            ++moved;
                         }
                     }
+                    if(moved == 0) {
+                        continue;
+                    }
                 }
             } else {
                 // from map/vehicle: add the item to the destination, if that worked,
                 // remove it from the source, else continue.
-                // TODO: move partial stack with items that are not counted by charges
-                item new_item( *sitem->it );
-                if( by_charges ) {
-                    new_item.charges = amount_to_move;
-                }
-                if( !add_item( destarea, new_item ) ) {
-                    // make sure to redraw the info window, so 
-                    // the user can see why it wasn't equipped
-                    if(destarea == AIM_WORN) {
-                        redraw = 1;
+                // create a new copy of the old item being manipulated
+                item new_item(*sitem->items.front());
+                // add the item, and note any items that might be leftover
+                int items_left = add_item(destarea, new_item, (by_charges) ? 1 : amount_to_move);
+                // only remove item or charges if the add succeeded
+                if(items_left == 0) {
+                    if(by_charges) {
+                        // set the new item's charge amount
+                        new_item.charges = amount_to_move;
+                        // `amount_to_move' will be `true' if the item needs to be removed
+                        amount_to_move = sitem->items.front()->reduce_charges(amount_to_move);
                     }
-                    continue;
+                    // take into account only the items _left over_ from transaction
+                    items_left -= remove_item(*sitem, amount_to_move);
                 }
-
-                if( by_charges && amount_to_move < sitem->it->charges ) {
-                    sitem->it->charges -= amount_to_move;
-                } else {
-                    remove_item( *sitem );
+                if(items_left > 0) {
+                    const char *msg = nullptr;
+                    if(by_charges) {
+                        msg = _("Only moved %d of %d charges.");
```

Comment:
If `by_charges`, the amount to move is always 1 (see the call to `add_item`), `items_left` can only be 0 or 1. The case 0 is handled in the other if-branch, so here items_left would 1 and *no* items or charges have been moved at all.

The message would read "Only moved 1 of 5439 charges." (assume I instructed to move 5439 charges). This is wrong (see comment below). Even if the number is fixed (`amount_to_move - items_left`), the message would read "Only moved 0 of 5439 charges." It should probably see "Move nothing" or so.